### PR TITLE
Create a common interface Transporter

### PR DESCRIPTION
**Pull request by hoenirvili**
Create a common interface to plug other forks to modify the http state
connection switching different underlying transport implementations.

### CODE REVIEW

**Comment by hoenirvili** [comment]
This patch enables client authentication with x509 certificates.

File: auth.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
+package winrm
+
+import (
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/masterzen/winrm/soap"
+	"launchpad.net/gwacl/fork/http"
```
Comment (masterzen) [inline auth.go]: You're still using the fork. At the minimum we should use the azure fork or better have this code run/compiled only on Go 1.7+

File: auth.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+		}
+	}
+
+	// make the custom transport
+	transport := &http.Transport{
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: endpoint.Insecure,
+			Certificates:       []tls.Certificate{cert},
+		},
+		//ResponseHeaderTimeout: endpoint.Timeout,
```
Comment (masterzen) [inline auth.go]: Why not use the timeout?

File: auth.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+// parse func reads the response body and return it as a string
+func parse(response *http.Response) (string, error) {
+
+	// if we recived the content we expected
+	if strings.Contains(response.Header.Get("Content-Type"), "application/soap+xml") {
+		body, err := ioutil.ReadAll(response.Body)
+		defer func() {
+			if err := response.Body.Close(); err != nil {
+				panic(err)
```
Comment (masterzen) [inline auth.go]: I'm not very fond of `panic` in libraries, couldn't it have a detrimental effect?

File: auth.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+			return "", fmt.Errorf("error while reading request body %s", err)
+		}
+
+		return string(body), nil
+	}
+
+	return "", fmt.Errorf("invalid content type")
+}
+
+func (c clientAuthRequest) Post(client *Client, request *soap.SoapMessage) (string, error) {
```
Comment (masterzen) [inline auth.go]: There's a lot of similarities between this code and the code in `http.go`. Isn't there a way to not have duplicate code at all (or at least minimize it)?

File: client.go [hunk 5]
```diff
@@ -31,23 +37,82 @@ func NewClient(endpoint *Endpoint, user, password string) (*Client, error) {
...
+		// if do you wish cu custom along more the transport
+		// this feature works very good with pluggble transports because
+		// the custom transports has the same types as in the stdlib
+		if params.TransportDecorator != nil && !t {
+			simpleRequest.transport = params.TransportDecorator(tr)
+		}
+
+		client.http = simpleRequest
+		// returun the client
+		return client, nil
```
Comment (masterzen) [inline client.go]: There's also a lot of code duplication between both cases of the switch. I'm wondering if there couldn't be a better abstraction that would end up in less duplication?

File: client.go [hunk 6]
```diff
@@ -85,7 +150,7 @@ func readCACerts(certs []byte) (*x509.CertPool, error) {
 
 // CreateShell will create a WinRM Shell,
 // which is the prealable for running commands.
-func (c *Client) CreateShell() (*Shell, error) {
+func (c Client) CreateShell() (*Shell, error) {
```
Comment (masterzen) [inline client.go]: You lost the pointer to `Client` again :)

File: client.go [hunk 7]
```diff
@@ -183,4 +248,4 @@ func (c Client) RunWithInput(command string, stdout, stderr io.Writer, stdin io.
 	cmd.Wait()
 
 	return cmd.ExitCode(), cmd.err
-}
+}
```
Comment (masterzen) [inline client.go]: I know this isn't really an issue anymore, but please leave the last `\n` in place (this is valid for the other files below), as at least it silences git and/or github diff.

File: client_test.go [hunk 8]
```diff
@@ -11,6 +11,14 @@ import (
 	. "gopkg.in/check.v1"
 )
 
+type Requester struct {
+	http func(*Client, *soap.SoapMessage) (string, error)
+}
+
+func (r Requester) Post(client *Client, request *soap.SoapMessage) (string, error) {
```
Comment (masterzen) [inline client_test.go]: Shouldn't `r` be `*Requester`?

File: client_test.go [hunk 9]
```diff
@@ -93,16 +103,16 @@ func (s *WinRMSuite) TestReplaceTransportWithDecorator(c *C) {
 
 	params := NewParameters("PT60S", "en-US", 153600)
 	params.TransportDecorator = func(*http.Transport) http.RoundTripper { return myrt }
-
+	_ = "breakpoint"
```
Comment (masterzen) [inline client_test.go]: You can safely remove this :)

File: auth.go (hunk 2 above)
Comment (hoenirvili) [inline auth.go]: On the gwacl version there were no ResponseHeaderTimeout attribute. But I will change to the azure fork.

File: auth.go (hunk 3 above)
Comment (hoenirvili) [inline auth.go]: I will remove the panic.

File: auth.go (hunk 4 above)
Comment (hoenirvili) [inline auth.go]: I will try to abstract it a little more.

File: client_test.go (hunk 8 above)
Comment (hoenirvili) [inline client_test.go]: No, we use *Requester if the method modifies the Requester state.

File: client_test.go (hunk 9 above)
Comment (hoenirvili) [inline client_test.go]: Fixed!

File: auth.go (hunk 1 above)
Comment (hoenirvili) [inline auth.go]: Added the Azure fork.

File: auth.go (hunk 3 above)
Comment (hoenirvili) [inline auth.go]: Fixed!

File: client.go (hunk 7 above)
Comment (hoenirvili) [inline client.go]: Fixed !

File: client.go (hunk 5 above)
Comment (hoenirvili) [inline client.go]: Fixed !

File: client.go (hunk 6 above)
Comment (hoenirvili) [inline client.go]: Fixed !

**Comment by hoenirvili** [comment]
Review?

File: client.go [hunk 10]
```diff
@@ -31,48 +36,49 @@ func NewClient(endpoint *Endpoint, user, password string) (*Client, error) {
...
-// newTransport will create a new HTTP Transport,
-// with options specified within the endpoint configuration
-func newTransport(endpoint *Endpoint) (*http.Transport, error) {
-
-	transport := &http.Transport{
-		TLSClientConfig: &tls.Config{
-			InsecureSkipVerify: endpoint.Insecure,
-		},
-		ResponseHeaderTimeout: endpoint.Timeout,
+		client.http = params.TransportDecorator()
```
Comment (bogdanteleaga) [inline client.go]: Isn't this supposed to be after the switch?

File: parameters.go [hunk 11]
```diff
@@ -1,16 +1,12 @@
...
-)
-
 // Parameters struct defines
 // metadata information and http transport config
 type Parameters struct {
 	Timeout            string
 	Locale             string
 	EnvelopeSize       int
-	TransportDecorator func(*http.Transport) http.RoundTripper
+	TransportDecorator func() Transporter
```
Comment (bogdanteleaga) [inline parameters.go]: This is more general indeed, but it will break users who depend on the library. It might now be possible to do it.

Comment (hoenirvili) [inline parameters.go]: @bogdanteleaga In the process of refactoring this code I was thinking on how to make this as transparent as possible without adding another parameter to NewClient or to NewClientWithParameters function so in my opinion this is a viable solution solution.

**Comment by hoenirvili** [comment]
@masterzen  merge?

**Comment by hoenirvili** [comment]
So what's the final verdict? @masterzen

**Comment by masterzen** [comment]
@hoenirvili I haven't had the time to review your proposed changes. I'm going to do that ASAP.

**Comment by hoenirvili** [comment]
@masterzen thanks for the response.

**Comment by hoenirvili** [comment]
busy day?

**Comment by hoenirvili** [comment]
@masterzen merge?

**Comment by hoenirvili** [comment]
@masterzen could you merge it already? been waiting some time and I need this patch to be merged upstream because I don't want to use/maintain my fork.

**Comment by masterzen** [comment]
I'm very very sorry @hoenirvili I haven't had the time to work on this project this last month. I'll do my best to review the code this week.

BTW, I think the TLS fix has been applied to Go 1.7 which is now in beta, I'm wondering if we couldn't wait for it to be released to remove the dependency on the Azure fork.

If you think we can't wait, I'll merge it as is once I've done the final review, we could still come back to this in a while.

Thanks!

**Comment by hoenirvili** [comment]
@masterzen I could refactor the code, but in the project that I will use this pkg is not supporting yet the latest patches of go ( go1.7), the project currently supports and uses 1.6.2 and It will take some years until it will change to the latest go version.

**Comment by hoenirvili** [comment]
@masterzen merge?

**Comment by masterzen** [comment]
@hoenirvili, I'm very sorry for the overdue review, I've been unfortunately busy by $work and $life things.

I've read the code again, and I still think there are lot of duplicated code like `auth.go parse` and `http.go body` or `auth.go Post` and `http.go Post`.

Isn't there a way were we could have those two functions in only one place and inject them their respective parameters?

What do you think?

**Comment by hoenirvili** [comment]
@masterzen I see your concern about the code duplication problem but really this is happening due to use of the  tls, http forks , because for example we can't make one type of another package with another type from another package work together even they are declared exactly the same as in naming, memory layout, etc, the compiler sees them differently and this is the prime reason for duplicating.

**Comment by masterzen** [comment]
@hoenirvili yes I understand. 

I'm wondering if we couldn't just have our own custom types that would delegate to the proper implementations when we call functions on them?

This might not be worth the effort, though.
Let me know what you think, thanks!

**Comment by hoenirvili** [comment]
@masterzen Also I would happly wait for the 1.7 go release and replace all the code to be just from the stlib , but the project I work and the dev-team only updates the go compiler/stdlib only on LTS(long time  releases).
So I must wait 4 years just to use this libary for the other project that I contribute to?

**Comment by hoenirvili** [comment]
@masterzen Is already done, I've created our own custom type Transporter and it's just doing exactly what you explained. It's choosing the underlying transport for the http connection and also the logic of the http post request of it's self.

```go
// switch to other transport if provided
	if params.TransportDecorator != nil {
		client.http = params.TransportDecorator()
	}

	// set the transport to some endpoint configuration
	if err := client.http.Transport(endpoint); err != nil {
		return nil, fmt.Errorf("Can't parse this key and certs: %s", err)
	}
```

If it's nothing special , it's choing the default transport, tls logic.

```go
client := &Client{
		Parameters: *params,
		username:   user,
		password:   password,
		url:        endpoint.url(),
		useHTTPS:   endpoint.HTTPS,
		// default transport
		http: &clientRequest{},
	}
```
The type ```clientRequest``` implements the ```Transporter Interface``` and it's the default code that is using **http**, **tls** from **go1.6.2** stdlib.

**Comment by hoenirvili** [comment]
@masterzen do you have time now to merge this?

**Comment by masterzen** [comment]
Let's merge it now.

**Pull request merged**